### ISSUE TITLE
`preprocessor_flags` vs. `compiler_flags`

### ISSUE THREAD

**Issue by zjturner** [facebook/buck2#203]
Why is there a distinction in the `cxx_library()` API between preprocessor flags and compiler flags?  I assumed that maybe by putting something in `preprocessor_flags`, buck2 would automatically prepend `-D`, but it turns out it doesn't do that.  It still just merges it with compiler flags as if they're identical.

When I think of preprocessor flags, I think of `-D` and `-U`, and not much else . But since it's just treating anything you put here as a regular command line argument, you could hypothetically put anything here.  You could put `-O3` here and it would "work", for example.  Where do I make the distinction?  Is an include path a preprocessor flag?   What about [/Zc:preprocessor](https://learn.microsoft.com/en-us/cpp/build/reference/zc-preprocessor?view=msvc-170) using MSVC?

I did observe that there is an `exported_preprocessor_flags` but no analogous `exported_compiler_flags`.  Why is this?  For example, our application builds using libc++ (not libstdc++), and we build using C++17.  So `--stdlib=libc++` and `--std=c++17` are flags we would like to pass unconditionally to every cpp file in our codebase.  If there were an `exported_compiler_flags` attribute this would be easy, but since there's not, the only workaround I can think of is to make a `my_cxx_library()` wrapper rule.  

Mostly curious about the design rationale behind these topics.

**Comment by zjturner** [comment]
Does this mean that I can use them interchangeably?  If that's the case, then I wonder if `preprocessor_flags` should just be preferred unconditionally, since it has the `exported_preprocessor_flags` version and `compiler_flags` doesn't, so it's strictly more flexible.

**Comment by ndmitchell** [comment]
Yes, you can mostly use them interchangeably, and for the reasons you list I would suggest preferring `preprocessor_flags`.

**Comment by cbarrete** [comment]
Would you accept a PR that adds `exported_compiler_flags`? I've just had to explain to a coworker that it was the right thing to do to avoid duplicating flags that should get inherited based on this conversation, and I'd like to make it less confusing.

**Comment by cbarrete** [comment]
Tagging @cjhopman and @JakobDegen, who I believe are familiar with this.
I have started looking into this, but I'm not sure which approach you'd prefer. The way I see it, we could either just add support for compiler flags in the exported preprocessor flags code (so more or less `s/exported_preprocessor/exported_preprocessor_and_compiler`), or duplicate all that logic and pretend that it is somewhat separate.
I favor the first one, but would like your opinion on what you would prefer.

**Comment by zjturner** [comment]
Wh as t do you mean by “avoid duplicating flags”?  Doesn’t it work to just use exported_preprocessor_flags and preprocessor_flags?

**Comment by cbarrete** [comment]
Yes it does, my users are just a little confused why they have to pass "regular" compiler flags as `exported_preprocessor_flags`. It's something that can be explained to every single dev working in the codebase, or wrapped away in a macro, or... well we can just add `exported_compiler_flags`, which is semantically correct.

**Comment by zjturner** [comment]
Ok, just making sure I understood.  I also wonder, what would break if you implemented it as:

```
ctx.attrs.exported_preprocessor_flags += ctx.attrs.exported_compiler_flags
```

at the beginning of the rule's impl.  Is there any semantic difference at all?

**Comment by cbarrete** [comment]
Not that I know of, and that's what I plan on doing if I can't get it in the prelude, but I don't see why we shouldn't have it in there though, it just makes sense to me to have consistent APIs for all users.

**Comment by cbarrete** [comment]
I had misunderstood your suggestion: I though you meant in the macro that I wrap prelude rules with.
Actually doing this in the prelude doesn't work because `ctx.attrs.exported_preprocessor_flags` is immutable by the time we get in the rule impl. Or at least, that's what I think is supposed to happen. When I try anyway, I get
```
Caused by:
    Traceback (most recent call last):
      File <builtin>, in <module>
      * build/prelude/rules.bzl:104, in buck2_compatibility_shim
          return impl(ctx)
    error: Operation `.exported_preprocessor_flags=` not supported on type `struct`
       --> build/prelude/cxx/cxx.bzl:182:5
        |
    182 |     ctx.attrs.exported_preprocessor_flags += ctx.attrs.exported_compiler_flags
        |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        |
```
